Read me (#51)
sebjshaw/Deloton-Project: README.md

- RESTful API
- Tableau integration to perform data querying and dashboard creation

## Project Roadmap

![Screenshot_Excalidraw](https://user-images.githubusercontent.com/115073814/217223449-26e6a315-d01c-4a11-a743-ceee6c4e60d7.png)

### Summary of above roadmap:

1. Kafka Consumer in EC2 takes logs from Kafka Data stream every second

> 2. Parses through logs and cleans data, passes into SQLite table ready for querying for Live Dashboard, displays information on current ride and recent rides

Reviewer: ```suggestion
2. Parses logs and cleans data, passes into SQLite table ready for querying from the Live Dashboard, displaying recent and current pages
```

> 3. At any point cleaned data suggests current user exceeds heart rate limit during ride, email alert trigger sent using Amazon simple email service (SES)

Reviewer: ```suggestion
3. At any point current user heart rate exceeds heart rate limit during the ride, calculated as a function of the user's age, email alert trigger sent using Amazon simple email service (SES)
```

Reviewer 2: ```suggestion
3. 3. Any time the current user's heart rate exceeds their heart rate limit during the ride, calculated as a function of their age, an email alert trigger is sent using Amazon simple email service (SES)
```

4. For a given ride for a specific user, at the end of their current ride, two csv files (user_info & ride_info) sent to s3 bucket (file storage service)

> 5. AWS Lambda function reads csv files from s3 bucket and extracts key metrics (using Pandas) for long term storage in AWS RDS postgres table

Reviewer: ```suggestion
5. AWS Lambda function reads csv files from s3 bucket and extracts key metrics (using Pandas) for long term storage in AWS RDS PostgreSQL table
```

6. RDS comprised of two tables, users and rides, joined by user_id
7. Tableau dashboard connected to RDS schemas and visuals created from it, live streaming the information

> 8. API hosted on EC2 performing SQL queries from RDS

Reviewer: ```suggestion
8. API hosted using AWS API Gateway, integrated with AWS Lambda Functions
```

9. Daily report hosted on AWS Lambda, extracting key metrics and formatted as an insightful summary, ran daily using Crontab

## Usage

### Live Dashboard hosted on EC2

### RESTful API

### Daily Report generator

### Heart Rate Email Alerts

1. EC2 hosting Kafka data stream automatically alerts user when their heart rate exceeds a certain level

> 2. Heart Rate Level calculated by age and weight

Reviewer: ```suggestion
2. Heart Rate Level calculated by age
```

## Usage

### Live Dashboard

1. You will find the live dashboard up and running here: [Live Dashboard](http://18.130.141.140:8080/)
2. Top left button allows user to switch between viewing the current ride (user details along with their current performance) and recent ride details.

#### Current Ride:
<img width="1411" alt="Screenshot_LiveDashboard" src="https://user-images.githubusercontent.com/115073814/217261046-79872c89-567d-4f6e-bf45-7550f1fbcbdf.png">

#### Recent Rides:
<img width="1414" alt="Screenshot_LiveDashboardRecent" src="https://user-images.githubusercontent.com/115073814/217261372-560e20da-d1ef-4a44-8d4f-6825283be560.png">

### RESTful API

[API](https://jzrx9wfk25.execute-api.eu-west-2.amazonaws.com/), can search the following endpoints:

```GET``` ```/ride/:id```
Get a ride with a specific ID

```GET``` ```/rider/:user_id```
Get rider information (e.g. name, gender, age, avg. heart rate, number of rides)

```GET``` ```/rider/:user_id/rides```
Get all rides for a rider with a specific ID

```DELETE``` ```/ride/:id```
Delete a with a specific ID

```GET``` ```/daily```
Get all of the rides in the current day

```GET``` ```/daily?date=01-01-2020```

> Get all rides for a specific date

Reviewer 2: nice addition of all the API endpoints

Author: I definitely didn't just nick that from the curriculum

- RESTful API
- Tableau integration to perform data querying and dashboard creation

## Project Roadmap

![Screenshot_Excalidraw](https://user-images.githubusercontent.com/115073814/217223449-26e6a315-d01c-4a11-a743-ceee6c4e60d7.png)

### Summary of above roadmap:

> 1. Kafka Consumer in EC2 takes logs from Kafka Data stream every second

Reviewer 2: ```suggestion
1. Kafka consumer polls the Deloton topic receiving two logs per second. Combines these logs into one containing all the data for that second of the ride
```

2. Following link to [Deloton-Project](https://prod-uk-a.online.tableau.com/t/threemusketeers/views/safe-copy/Dashboard-RidesCompleted) prompts a login with AWS RDS credentials. Postgres account login details provided on day of presentation.
3. Dashboard consists of eight pages, each of which a leaderboard for the top users for that measured metric i.e. Number of rides completed by user, Highest average heart rate per ride per user.
4. User may navigate via the buttons on the righthand side of each page
5. Filters by gender, minimum & maximum value, and date of ride for interactability

![Screenshot_Tableau](https://user-images.githubusercontent.com/115073814/217216401-ccf2e62f-8258-485f-9d56-18a5df5a5859.png)

## Credits and Acknowledgements

Direct contributors to the repository:
- Seb Shaw: [sebjshaw](https://github.com/sebjshaw)
- Dominic Lawson: [DomLaw82](https://github.com/DomLaw82)
- Alexander Skowronski: [AlexSkowronski2](https://github.com/AlexSkowronski2)

Resources used:
- [Pandas](https://pandas.pydata.org/)
- [Amazon Web Services](https://aws.amazon.com/)
    - API Gateway
    - EC2
    - ECR
    - Lambda
    - S3

>     - RDS

Reviewer 2: ```suggestion
    - RDS
    - SES
```
